tests/scripts/thread-cert/border_router/nat64/test_upstream_dns.py

```diff
@@ -0,0 +1,133 @@
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+import unittest
+
+import config
+import thread_cert
+
+import ipaddress
+import shlex
+
+# Test description:
+#   This test verifies forwarding DNS queries sent by 'Router' by using
+# a record resolved by BIND9 server.
+#
+# Topology:
+#    ----------------(eth)--------------------
+#           |                 |
+#          BR (Leader)      HOST
+#           |
+#        ROUTER
+#
+
+BR = 1
+ROUTER = 2
+HOST = 3
+
+TEST_DOMAIN = 'test.domain'
+TEST_DOMAIN_IP6_ADDRESSES = {'2001:db8::1'}
+
+
+class UpstreamDns(thread_cert.TestCase):
+    USE_MESSAGE_FACTORY = False
+
+    TOPOLOGY = {
+        BR: {
+            'name': 'BR',
+            'allowlist': [ROUTER],
+            'is_otbr': True,
+            'version': '1.3',
+        },
+        ROUTER: {
+            'name': 'Router',
+            'allowlist': [BR],
+            'version': '1.3',
+        },
+        HOST: {
+            'name': 'Host',
+            'is_host': True
+        },
+    }
+
+    def test(self):
+        br = self.nodes[BR]
+        router = self.nodes[ROUTER]
+        host = self.nodes[HOST]
+
+        host.start(start_radvd=False)
+        self.simulator.go(5)
+
+        br.start()
+        # When feature flag is enabled, NAT64 might be disabled by default. So
+        # ensure NAT64 is enabled here.
+        self.simulator.go(config.LEADER_STARTUP_DELAY)
+        self.assertEqual('leader', br.get_state())
+
+        br.nat64_set_enabled(True)
+        br.srp_server_set_enabled(True)
+
+        br.bash('service bind9 stop')
+
+        br.bash(
+            shlex.join(['echo', f'zone "{TEST_DOMAIN}" {{ type master; file "/etc/bind/db.test.domain"; }};']) +
+            ' >> /etc/bind/named.conf.local')
+        br.bash(shlex.join(['echo', '$TTL 24h']) + ' >> /etc/bind/db.test.domain')
+        br.bash(
+            shlex.join(['echo', f'@ IN SOA {TEST_DOMAIN} test.{TEST_DOMAIN}. ( 20230317 86400 300 604800 3600 )']) +
+            ' >> /etc/bind/db.test.domain')
+        br.bash(shlex.join(['echo', f'@ IN NS {TEST_DOMAIN}.']) + ' >> /etc/bind/db.test.domain')
+        for addr in TEST_DOMAIN_IP6_ADDRESSES:
+            br.bash(shlex.join(['echo', f'@ IN AAAA {addr};']) + ' >> /etc/bind/db.test.domain')
```

Comment:
Could we wrap these into a function for better readability?

Reply:
extracted this into a single string block, hope it has better readability now